source/fuzz/transformation_merge_function_returns.cpp

```diff
@@ -0,0 +1,646 @@
+  message_.set_function_id(function_id);
+  message_.set_outer_header_id(outer_header_id);
+  message_.set_outer_return_id(outer_return_id);
+  message_.set_return_val_id(return_val_id);
+  message_.set_any_returnable_val_id(any_returnable_val_id);
+  for (const auto& return_merging_info : returns_merging_info) {
+    *message_.add_return_merging_info() = return_merging_info;
+  }
+}
+
+bool TransformationMergeFunctionReturns::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  auto function = ir_context->GetFunction(message_.function_id());
+  // The function must exist.
+  if (!function) {
+    return false;
+  }
+
+  // The entry block must end in an unconditional branch.
+  if (function->entry()->terminator()->opcode() != SpvOpBranch) {
+    return false;
+  }
+
+  // If the function has a non-void return type,
+  // |message_.any_returnable_val_id| must exist, have the same type as the
+  // return type of the function and be available at the end of the entry block.
+  auto function_type = ir_context->get_type_mgr()->GetType(function->type_id());
+  assert(function_type && "The function type should always exist.");
+
+  // Get a map from the types for which ids are available at the end of the
+  // entry block to one of the ids with that type. We compute this here to avoid
+  // potentially doing it multiple times later on.
+  auto types_to_available_ids =
+      GetTypesToIdAvailableAfterEntryBlock(ir_context);
+
+  if (!function_type->AsVoid()) {
+    auto returnable_val_def =
+        ir_context->get_def_use_mgr()->GetDef(message_.any_returnable_val_id());
+    if (!returnable_val_def) {
+      // Check if a suitable id can be found in the module.
+      if (types_to_available_ids.count(function->type_id()) == 0) {
+        return false;
+      }
+    } else if (returnable_val_def->type_id() != function->type_id()) {
+      return false;
+    } else if (!fuzzerutil::IdIsAvailableBeforeInstruction(
+                   ir_context, function->entry()->terminator(),
+                   message_.any_returnable_val_id())) {
+      // The id must be available at the end of the entry block.
+      return false;
+    }
+  }
+
+  // Get the reachable return blocks.
+  auto return_blocks =
+      fuzzerutil::GetReachableReturnBlocks(ir_context, message_.function_id());
+
+  // Get all the merge blocks of loops containing reachable return blocks.
+  std::set<uint32_t> merge_blocks;
+  for (uint32_t block : return_blocks) {
+    uint32_t merge_block =
+        ir_context->GetStructuredCFGAnalysis()->LoopMergeBlock(block);
+    while (merge_block != 0 && !merge_blocks.count(merge_block)) {
+      merge_blocks.emplace(merge_block);
+      merge_block =
+          ir_context->GetStructuredCFGAnalysis()->LoopMergeBlock(merge_block);
+    }
+  }
+
+  // All of the relevant merge blocks must not contain instructions whose opcode
+  // is not one of OpLabel, OpPhi or OpBranch.
```

Comment:
I would avoid double-negatives, and say:

"Instructions in the relevant merge blocks must be restricted to OpLabel, OpPhi and OpBranch."

I suggest rephasing the similar comment in the .proto file similarly.

Reply:
Done